File: InterviewBit/Linked LIst, Stack & Queues/Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
ListNode* Solution::mergeTwoLists(ListNode* A, ListNode* B) {
    if(A==NULL) return B;
    if(B==NULL) return A;

    ListNode *root, *k;
    root=new ListNode(0);
    k=root;
    while(A!=NULL && B!=NULL){
        if(A->val < B->val){
            root->next = new ListNode(A->val);
            A=A->next;
            root=root->next;
        }else{
            root->next = new ListNode(B->val);
            B=B->next;
            root=root->next;
        }
    }

    while(A!=NULL){
        root->next = new ListNode(A->val);
        A=A->next;
        root=root->next;
    }

    while(B!=NULL){
        root->next = new ListNode(B->val);
        B=B->next;
        root=root->next;
    }

    return k->next;
}

File: LeetCode/Merge-Two-Sorted-Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null)
            return list2;
        if (list2 == null)
            return list1;

        ListNode temp = null;

        if (list1.val <= list2.val) {
            temp = list1;
            list1 = list1.next;
        } else {
            temp = list2;
            list2 = list2.next;
        }

        ListNode resp = temp;

        while (list1 != null || list2 != null) {
            if (list1 != null && list2 != null) {
                if (list1.val <= list2.val) {
                    temp.next = list1;
                    list1 = list1.next;
                } else {
                    temp.next = list2;
                    list2 = list2.next;
                }
                temp = temp.next;
            } else if (list1 != null) {
                temp.next = list1;
                return resp;
            } else {
                temp.next = list2;
                return resp;
            }
        }

        return resp;
    }
}